### ISSUE TITLE
Container.forceReadOnly(true) sometimes closes the Container

### ISSUE THREAD

**Issue by scottn12** [microsoft/FluidFramework#10191]
## Describe the bug
There is an intermittent issue where calling `Container.forceReadOnly(true)` will sometimes cause the container to unexpectedly close. Additionally, a `fluid:telemetry:ConnectionManager ForceReadonlyPendingChanged` error is thrown in this scenario as well.

### To Reproduce
1. Create a container
2. The container is connected
3. The container is made dirty
4. Wait for `container.on("saved")` to fire and ensure `container.isDirty === false`
5. Call `container.forceReadOnly(true)`

Note: This can be difficult to reproduce in certain environments. For help reproducing, reach out to @bensonms and @kmui.

### Expected behavior
`Container.forceReadOnly(true)` should not cause the container to close.

### Logs
```
[Error] fluid:telemetry:ConnectionManager ForceReadonlyPendingChanged {"category":"error","clientType":"interactive","containerId":"dfb4fc3a-f3b9-4591-ab9a-fe341bfa708c","docId":"8SRmVQ2fz9nQeDbYV13gujNP5PX1iro9S5He84WPKnw[object Object]","containerAttachState":"Attached","containerLifecycleState":"loaded","containerConnectionState":"Connected","dmInitialSeqNumber":0,"dmLastKnownSeqNumber":12,"dmLastMsqSeqNumber":11,"dmLastMsqSeqTimestamp":1651703422621,"dmLastMsqSeqClientId":"4a3adbd3-fb53-4b1d-a8d4-6b8e623c7b3c","connectionState":"Connected","connectionStateDuration":94225,"loaderId":"a0fee9e8-461c-4ca7-8a00-91a74c6798d9","loaderVersion":"0.54.3"} tick=97016 %6c +0ms
[Error] fluid:telemetry:Container ContainerClose {"category":"error","error":"0x1d1","errorType":"dataProcessingError","errorInstanceId":"7e9b7770-747b-41a3-ae5e-58fcd7ae04f7","fluidErrorCode":"","canRetry":false,"untrustedOrigin":1,"dataProcessingError":1,"dataProcessingCodepath":"deltaManagerInboundErrorHandler","messageClientId":"4a3adbd3-fb53-4b1d-a8d4-6b8e623c7b3c","messageSequenceNumber":12,"messageClientSequenceNumber":11,"messageReferenceSequenceNumber":1,"messageMinimumSequenceNumber":1,"messageTimestamp":1651703422621,"message":"0x1d1","clientType":"interactive","containerId":"dfb4fc3a-f3b9-4591-ab9a-fe341bfa708c","docId":"8SRmVQ2fz9nQeDbYV13gujNP5PX1iro9S5He84WPKnw[object Object]","containerAttachState":"Attached","containerLifecycleState":"closing","containerConnectionState":"Disconnected","dmInitialSeqNumber":0,"dmLastKnownSeqNumber":12,"dmLastMsqSeqNumber":12,"dmLastMsqSeqTimestamp":1651703422621,"dmLastMsqSeqClientId":"4a3adbd3-fb53-4b1d-a8d4-6b8e623c7b3c","connectionState":"Disconnected","connectionStateDuration":65512,"loaderId":"a0fee9e8-461c-4ca7-8a00-91a74c6798d9","loaderVersion":"0.54.3"} tick=162530 Error
```

### Additional Information
Using FluidFramework version 0.58.2002

**Comment by scottn12** [comment]
@vladsud, would you be able to take a look at this? I'm having trouble finding the root cause.

**Comment by markfields** [comment]
Quick update from Teams thread - @anthony-murphy had an interesting theory about the container returning !dirty, but there could still be an outstanding code proposal.

The container closed due to assert 0x1d1 (local code proposal not tracked properly, if I'm reading it right) being thrown through the `deltaManagerInboundErrorHandler` code path.

**Comment by curtisman** [comment]
@vladsud I think you might be right in terms Quorum behavior. 

A question though is that if we are in the process of processing multiples ops (may be batched), and force disconnect happens in the middle, is it expected that the runtime is in a consistent state to continue to process those ops (i.e. process incoming ops while disconnected).  I would guess that most op processing code doesn't take that into account.

**Comment by ssimic2** [comment]
@vladsud maybe help us clarify this: "proposals that were already sent will be acked by service, and suck ops will be processed as local at a later connection time". I'm assuming ack here is our own sequenced proposal op... If client reconnects it gets a new client id, meaning even if subsequent proposal ack op came in, it would not be treated as local ?

I'm asking because of short-term mitigation (FFX are integrating with our release this week). Our assumption here is that this error can **only** occur when there are ops on the inbound queue at the time disconnect happens. Only in that case `queueMicrotask()` would make sense, so wanted to confirm there are no other scenarios we are missing.

Me and @scottn12 are OCEs. We are looking at mitigations needed for FFX integration, and we'll get Matt's help here.

**Comment by vladsud** [comment]
clientID is switched to new one (on reconnection) after loader ensured there is 0% chance for ops sent by this session to land in a file. In other words, connection stays in pending state (and looks like disconnected to runtime) while loader waits for acknowledgement of all previously in-flight ops to either be acked and processed or lost forever.
This is required to ensure that DDSs resubmit any pending changes only if they were not acked, i.e. avoid double send.

All such acked local ops will show up as "local" while we are in this transition state - they were sent by this client, and DDSs have to properly account them as local and ensure DDSs will not resubmit same ops again once we are done with connection dance.

This has nothing to do with inbound queue. Ops can be in transition for minutes.
We use combination of new join op and old leave op as barrier to know when it's safe to switch to new clientID and report back ot runtime that connection occurs. Till that barrier we operate with old clientId.

**Comment by vladsud** [comment]
@curtisman, as we chatted, forceReadonly can be improved to not disconnect in the middle of the batch. But I'd rather treat it as separate improvement (with its own assessment of risks of this change). Quorum is broken without it as disconnect can happen at any time, and quorum ops are never part of a batch.

**Comment by ChumpChief** [comment]
@vladsud I was noting that you added the `localProposals.clear()` a long time ago with this note [in this change](https://github.com/microsoft/FluidFramework/commit/ecb96c8dfc77ea8f8848bec97226bdf624f048c7), and was wondering if you remember the context (i.e. is there some known bug that it fixed, which we should be cautious not to regress with a change here):

> * Propagating connected state to quorum and properly handling leader election & rejection of local proposals on disconnect.
Before this change we
>     - would make leadership proposal in Connecting state, causing it to be dropped and not elect a leader
>     - any proposals that made it to DeltaManager (but not necessarily to server) would hang in the air on disconnect.

I am not proposing to resubmit -- your prior comment suggested that it should resubmit non-acked proposals and I'm pointing out that we can't do that safely:

> Quorum violates how DDSs should work - it should not clear local proposals, and instead resubmit them on reconnection (unless they are acked on reconnection).

In any case, I'm looking into it.

**Comment by ChumpChief** [comment]
Met with FFX, a couple notes:

Directly relevant to the FFX issue:
- I've got a tentative Quorum change that would permit proposal promises to linger unsettled until reconnect, and decide whether they are resolved/rejected then.  Looks ok in initial testing, will beat on it a bit more and put up a PR.  This will eliminate the 1d1 error they were seeing, but this will not stop the ForceReadonlyPendingChanged error.
- The workaround to use `queueMicrotask()` to defer the call to `forceReadnly()` seems to alleviate the ForceReadonlyPendingChanged error in this case.  It is not sufficient as a general solution as there's not a strong guarantee that the proposal op will be sitting in the inbound queue at the moment the container is "saved".
- Meeting with FFX, we discovered that the code proposal is probably unintentional/undesired (the expected code details are not stamped during `createDetachedContainer()` and so this proposal is to "fix up" a newly created container).  Optimally, there would be no code proposal in this flow.  FFX will investigate.
- Another couple approaches I thought about after the meeting:
    - The scenario wants to be readonly from the start -- there at least used to be a way to load a container in readonly state?  Not sure what the latest on that is.  Probably not a viable approach unless/until they stop writing the extra code proposal though.
    - The FF concept of readonly might not be the best tool for this scenario.  Since it's invalid to enter data at this point in the scenario, it might be better to defer even creating the container until it's OK to enter data (or, create it but keep it detached).

Not immediately relevant to the FFX issue but worth considering:
- `ContainerRuntime` calls `updateDocumentDirtyState(false)` synchronously from the `processDeltas()` stack as soon as the last outstanding local op is seen.
    - From one perspective this is ok -- this is the earliest that we know all of our sent ops have been ack'd.  This is totally sufficient if this state is just being used to show/hide banner UI ("saved means the server has seen all your local ops").
    - From another perspective it's problematic -- we are "saved" before the op gets fully processed -- i.e. before `this.dataStores.processFluidDataStoreOp()` gets called with our last local op.  This is problematic if the event is used to to infer about the actual state of the data, since synchronous access of the data from a "saved" handler will not see that last op applied.
        - We might consider moving the "saved" event to fire after that local op is fully processed ("saved means your data has resolved the acks of your local ops")
- `forceReadonly()` needs to clarify its contract with:
    - "saved"/"dirty" state - is it a requirement for callers to verify the document is "saved" before calling?
    - connection state - what's the expectation if `forceReadonly()` is called during connecting/disconnected state?

**Comment by vladsud** [comment]
I do not remember much from 3 years ago :) I think this code was not dealing with reconnects at all, so it was my attempt to add this missing feature.

I'd suggest clearing local proposals on connection, not disconnection - seems more correct (though pending proposals will not be cleared on container close).

**Comment by anthony-murphy** [comment]
for createDetachedContainer() there was talk in the past of changing what that accepts. it's probably something like a code details, but the idea was to let the code loader, which actually provides the runtime factory give the real code details that is stamped int he quorum. Then the code details in createDetachedContainer() is used for things like determining which container to use, or parameters for the container (blueprint). I think this makes sense, as the code loader really knows the truth and its a pain to orchestrate that out of the code loader to the host just to pass it back in

**Comment by ChumpChief** [comment]
>for createDetachedContainer() there was talk in the past of changing what that accepts. it's probably something like a code details, but the idea was to let the code loader, which actually provides the runtime factory give the real code details that is stamped in the quorum. Then the code details in createDetachedContainer() is used for things like determining which container to use, or parameters for the container (blueprint). I think this makes sense, as the code loader really knows the truth and its a pain to orchestrate that out of the code loader to the host just to pass it back in

I think we should have a design discussion if we consider a change like that.  If the host is passing in a code loader + "not exactly code details but I trust that my code loader can figure it out", then it implies to me that the host must have knowledge that there's a contract between the code loader and those not-details.

In that case they could directly ask the code loader to resolve to actual details before calling `createDetachedContainer()` (rather than adding to the code loader interface and adding a responsibility to the Container).  This host-to-code-loader interaction would be more parallel to `proposeCodeIfNewer()`-type flows too, where the host (who is the expected caller of `proposeCodeDetails()`) ideally knows whether a call to `proposeCodeDetails()` is appropriate, should have an effect vs. no-op, will force all clients to reload, etc. before making the call.

Moreover I don't know that the code loader should really hold that detail-resolving responsibility either - if the scenario is something like "user picks a container type from a library" then perhaps there's some library/catalog that should be providing back the fully-resolved details instead, leaving the code loader focused solely on loading.

**Comment by anthony-murphy** [comment]
The above idea is more around the ergonomics of implementation, and agree it needs design if it's a path we'd want to follow, which is why it was never done :)  We could also think about it more of a pattern, than something we'd concretely implement, as it does imply a tight coupling between the host, code loader, and container.

**Comment by ChumpChief** [comment]
Closing this issue -- results are:

1. 0x1d1 error is fixed #10213 
2. Superfluous proposal for Loop tracked by https://office.visualstudio.com/OC/_workitems/edit/6042550/
3. I'm tracking the issue of container code being unaware of code proposal as part of #8176, along with any other ideas about changes to code proposal.
4. I think the forceReadonly() in non-connected states is sufficiently tracked by #9787.

And I think that's all we really need to track here, lmk if I missed anything.

**Comment by markfields** [comment]
There's another repro here:  AB#1246
